### PR TITLE
Update function type for return with multiple operands, some repeating

### LINKED ISSUE #1350: Handle updating all function result types in --stablehlo-refine-shapes
### What happened?

Shape refinement fails to update function result type when updating return value.

Unsure if this is exclusive to multiple outputs/inputs, see repro, looks like the first result type gets updated, not the second.

### Steps to reproduce your issue

Example testpoint:

```
// RUN: stablehlo-opt --stablehlo-refine-shapes %s
module @jit_fun_flat_jax {
  func.func @main(%arg0: tensor<3x2xf32>, %arg1: tensor<3x2xf32>, %arg2: tensor<i32>) -> (tensor<?x2xf32>, tensor<i1>, tensor<?x2xf32>, tensor<i32>) {
    %0 = stablehlo.constant dense<1> : tensor<i32>
    %1 = stablehlo.constant dense<2> : tensor<i32>
    %2 = stablehlo.constant dense<0> : tensor<i32>
    %3 = stablehlo.add %arg2, %0 : tensor<i32>
    %4 = stablehlo.add %arg1, %arg0 : (tensor<3x2xf32>, tensor<3x2xf32>) -> tensor<?x2xf32>
    %5 = stablehlo.convert %1 : tensor<i32>
    %6 = stablehlo.compare  EQ, %5, %2,  SIGNED : (tensor<i32>, tensor<i32>) -> tensor<i1>
    %7 = stablehlo.select %6, %0, %5 : tensor<i1>, tensor<i32>
    %8 = stablehlo.remainder %3, %7 : tensor<i32>
    %9 = stablehlo.compare  NE, %8, %2,  SIGNED : (tensor<i32>, tensor<i32>) -> tensor<i1>
    %10 = stablehlo.compare  LT, %8, %2,  SIGNED : (tensor<i32>, tensor<i32>) -> tensor<i1>
    %11 = stablehlo.compare  LT, %7, %2,  SIGNED : (tensor<i32>, tensor<i32>) -> tensor<i1>
    %12 = stablehlo.compare  NE, %10, %11,  UNSIGNED : (tensor<i1>, tensor<i1>) -> tensor<i1>
    %13 = stablehlo.and %12, %9 : tensor<i1>
    %14 = stablehlo.add %8, %7 : tensor<i32>
    %15 = stablehlo.select %13, %14, %8 : tensor<i1>, tensor<i32>
    %16 = stablehlo.compare  EQ, %15, %2,  SIGNED : (tensor<i32>, tensor<i32>) -> tensor<i1>
    return %4, %16, %4, %3 : tensor<?x2xf32>, tensor<i1>, tensor<?x2xf32>, tensor<i32>
  }
}
```

Errors with the following invalid IR and error:

```
"func.func"() ({
^bb0(%arg0: tensor<3x2xf32>, %arg1: tensor<3x2xf32>, %arg2: tensor<i32>):
  %0 = "stablehlo.constant"() {value = dense<false> : tensor<i1>} : () -> tensor<i1>
  %1 = "stablehlo.constant"() {value = dense<2> : tensor<i32>} : () -> tensor<i32>
  %2 = "stablehlo.constant"() {value = dense<1> : tensor<i32>} : () -> tensor<i32>
  %3 = "stablehlo.constant"() {value = dense<0> : tensor<i32>} : () -> tensor<i32>
  %4 = "stablehlo.add"(%arg2, %2) : (tensor<i32>, tensor<i32>) -> tensor<i32>
  %5 = "stablehlo.add"(%arg1, %arg0) : (tensor<3x2xf32>, tensor<3x2xf32>) -> tensor<3x2xf32>
  %6 = "stablehlo.remainder"(%4, %1) : (tensor<i32>, tensor<i32>) -> tensor<i32>
  %7 = "stablehlo.compare"(%6, %3) {compare_type = #stablehlo<comparison_type SIGNED>, comparison_direction = #stablehlo<comparison_direction NE>} : (tensor<i32>, tensor<i32>) -> tensor<i1>
  %8 = "stablehlo.compare"(%6, %3) {compare_type = #stablehlo<comparison_type SIGNED>, comparison_direction = #stablehlo<comparison_direction LT>} : (tensor<i32>, tensor<i32>) -> tensor<i1>
  %9 = "stablehlo.compare"(%8, %0) {compare_type = #stablehlo<comparison_type UNSIGNED>, comparison_direction = #stablehlo<comparison_direction NE>} : (tensor<i1>, tensor<i1>) -> tensor<i1>
  %10 = "stablehlo.and"(%9, %7) : (tensor<i1>, tensor<i1>) -> tensor<i1>
  %11 = "stablehlo.add"(%6, %1) : (tensor<i32>, tensor<i32>) -> tensor<i32>
  %12 = "stablehlo.select"(%10, %11, %6) : (tensor<i1>, tensor<i32>, tensor<i32>) -> tensor<i32>
  %13 = "stablehlo.compare"(%12, %3) {compare_type = #stablehlo<comparison_type SIGNED>, comparison_direction = #stablehlo<comparison_direction EQ>} : (tensor<i32>, tensor<i32>) -> tensor<i1>
  "func.return"(%5, %13, %5, %4) : (tensor<3x2xf32>, tensor<i1>, tensor<3x2xf32>, tensor<i32>) -> ()
}) {function_type = (tensor<3x2xf32>, tensor<3x2xf32>, tensor<i32>) -> (tensor<3x2xf32>, tensor<i1>, tensor<?x2xf32>, tensor<i32>), sym_name = "main"} : () -> ()

...

error: type of return operand 2 ('tensor<3x2xf32>') doesn't match function result type ('tensor<?x2xf32>') in function @main
    return %4, %16, %4, %3 : tensor<?x2xf32>, tensor<i1>, tensor<?x2xf32>, tensor<i32>
    ^
/tmp/sr.mlir:21:5: note: see current operation: "func.return"(%5, %13, %5, %4) : (tensor<3x2xf32>, tensor<i1>, tensor<3x2xf32>, tensor<i32>) -> ()
```

### Version information

_No response_

### PR DESCRIPTION
**Pull request by GleasonK**
The way the refine function type algorithm works currently has an assumption that ReturnOp does not have redundant operands. This is not always the case in practice.

Several JAX tests were hitting this issue.

Closes #1350

### CODE REVIEW

**Comment by GleasonK** [comment]
cc @gnecula

File: stablehlo/tests/stablehlo_refine_shapes.mlir [hunk 1]
```diff
@@ -500,6 +500,15 @@ func.func @refine_dynamic_reshape(%arg0: tensor<4xf32>) -> tensor<*xf32> {
 
 // -----
 
+// Test for #1350
+// CHECK-LABEL: @refine_function_type
+func.func @refine_function_type(%arg0: tensor<3x2xf32>, %arg1: tensor<3x2xf32>) -> (tensor<3x2xf32>, tensor<3x2xf32>) {
+  %0 = stablehlo.add %arg0, %arg1 : tensor<3x2xf32>
+  return %0, %0 : tensor<3x2xf32>, tensor<3x2xf32>
+}
```
Comment (burmako) [inline stablehlo/tests/stablehlo_refine_shapes.mlir:508]: Does this test actually exercise the new behavior? Let's have a test, similarly to other tests nearby which has static arguments but unranked results, and then refines all the results to be static?

Comment (GleasonK) [reply stablehlo/tests/stablehlo_refine_shapes.mlir:508]: Good catch, I coped the output of my test case, not the test itself. Updated so output types all have dynamic shapes.

Comment (GleasonK) [reply stablehlo/tests/stablehlo_refine_shapes.mlir]: Moved to `update_function_type_multiple_outputs`

**Review by burmako — APPROVED** [review]

**Pull request merged**